File: 01_slidingWindow/longestSubstr.cpp
using namespace std;

#include <iostream>
#include <string>
#include <vector>

class NoRepeatSubstring {
 public:
  static int findLength(const string& str) {
    int maxLength = 0;
    vector<int> dict(128,-1);
    int startSt=0;
    for(int endSt=0;endSt<str.length();++endSt){
      if(dict[str[endSt]]>-1){
        startSt=dict[str[endSt]]+1;
      }
      dict[str[endSt]]=endSt;
      maxLength=max(maxLength,endSt-startSt+1);
    }
    return maxLength;
  }
};

int main(int argc, char* argv[]) {
  cout << "Length of the longest substring: " << NoRepeatSubstring::findLength("aabccbb") << endl;
  cout << "Length of the longest substring: " << NoRepeatSubstring::findLength("abbbb") << endl;
  cout << "Length of the longest substring: " << NoRepeatSubstring::findLength("abccde") << endl;
  cout << "Length of the longest substring: " << NoRepeatSubstring::findLength("bbabcabcdd") << endl;
}

File: 01_slidingWindow/longestSubstr.py
def non_repeat_substring(str1):
  start_point = 0
  max_length = 0
  char_index_map = {}

  for end_point in range(len(str1)):
    right_char = str1[end_point]
    if right_char in char_index_map:
      start_point = max(start_point, char_index_map[right_char] + 1)
    char_index_map[right_char] = end_point
    max_length = max(max_length, end_point - start_point + 1)
  return max_length


def main():
  print("Length of the longest substring: " + str(non_repeat_substring("aabccbb")))
  print("Length of the longest substring: " + str(non_repeat_substring("abbbb")))
  print("Length of the longest substring: " + str(non_repeat_substring("abccde")))
  print("Length of the longest substring: " + str(non_repeat_substring("bbabcabcdd")))


main()